counsel.el

```diff
@@ -4517,6 +4517,71 @@ If there is no such buffer, start a new `shell' with NAME."
                             (reusable-frames . visible)))
     (shell name)))
 
+;;** `counsel-firefox-bookmarks'
+(defvar counsel-firefox-bookmarks-file
+  (car (file-expand-wildcards "~/.mozilla/firefox/*/bookmarks.html"))
+  "Firefox's automatically exported HTML bookmarks file.")
+
+(defface counsel-firefox-bookmarks-tag
+  '((t :inherit font-lock-comment-face))
+  "Face used by `counsel-firefox-bookmarks' for tag."
+  :group 'ivy-faces)
+
+(defun counsel-firefox-bookmarks-action (x)
+  "Browse candidate X."
+  (browse-url (cdr x)))
+
+(declare-function xml-substitute-special "xml")
+
+(defun counsel-firefox-bookmarks--candidates ()
+  "Return a list of completion candidates for `counsel-firefox-bookmarks'."
+  (unless (and counsel-firefox-bookmarks-file
+               (file-readable-p counsel-firefox-bookmarks-file))
+    (signal 'file-error (list "Opening `counsel-firefox-bookmarks-file'"
+                              "No such readable file"
+                              counsel-firefox-bookmarks-file)))
+  (require 'xml)
+  (with-temp-buffer
+    (insert-file-contents counsel-firefox-bookmarks-file)
+    (let ((case-fold-search t)
+          candidates)
+      (while (re-search-forward
+              "<a href=\"\\([^\"]+?\\)\"[^>]*?>\\([^<]*?\\)</a>" nil t)
+        (let* ((a (match-string 0))
+               (href (match-string 1))
+               (text (if (= (match-beginning 2) (match-end 2))
+                         href
+                       (save-match-data
+                         (xml-substitute-special (match-string 2)))))
+               (tags (and (string-match "tags=\"\\([^\"]+?\\)\"" a)
+                          (match-string 1 a))))
+          (push (cons (if tags
+                          (concat text
+                                  " :"
+                                  (string-join
```

Comment:
`string-join` was only added in Emacs 24.4, so Counsel does not support it. If Counsel supported it, you would have to `(eval-when-compile (require 'subr-x))` in order to use it.